jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java

```diff
@@ -0,0 +1,225 @@
+     * @param perfectEliminationOrder the perfect elimination order of the graph
+     * @return a nice decomposition builder for the graph if the graph was chordal, else null
+     */
+    public static <V, E> ChordalNiceDecompositionBuilder<V, E> create(
+        Graph<V, E> graph, List<V> perfectEliminationOrder)
+    {
+        ChordalNiceDecompositionBuilder<V, E> builder =
+            new ChordalNiceDecompositionBuilder<>(graph, perfectEliminationOrder);
+        builder.computeNiceDecomposition();
+        return builder;
+
+    }
+
+    /**
+     * Creates a nice decomposition builder for chordal graphs.
+     * 
+     * @param graph the chordal graph
+     * @param perfectOrder the perfect elimination order of graph
+     */
+    private ChordalNiceDecompositionBuilder(Graph<V, E> graph, List<V> perfectOrder)
+    {
+        super();
+        this.graph = graph;
+        this.perfectOrder = perfectOrder;
+        vertexInOrder = getVertexInOrder();
+    }
+
+    /**
+     * Computes the nice decomposition of the graph. We iterate over the perfect elimination order
+     * of the chordal graph and try to add a node containing the predecessors regarding the
+     * perfect elimination as a bag to the tree
+     */
+    private void computeNiceDecomposition()
+    {
+
+        // map from vertices to decomposition-nodes where the decomposition-node has all its
+        // predecessors
+        Map<V, Integer> forgetNodeMap = new HashMap<V, Integer>(graph.vertexSet().size());
+
+        // set current node to the root
+        Integer decompNode = getRoot();
+
+        // iterate over the perfect elimination order
+        for (V vertex : perfectOrder) {
+
+            // get the predecessors regarding the perfect elimination order
+            Set<V> predecessors = getOrderPredecessors(vertexInOrder, vertex);
+
+            // calculate nearest successors according to perfect elimination order
+            V lastVertex = null;
+            for (V predecessor : predecessors) {
+                if (lastVertex == null)
+                    lastVertex = predecessor;
+                if (vertexInOrder.get(predecessor) > vertexInOrder.get(lastVertex))
+                    lastVertex = predecessor;
+            }
+
+            // get node with clique of last vertex, else we use the last node
+            if (lastVertex != null)
+                decompNode = forgetNodeMap.get(lastVertex);
+
+            // if this node is not a leaf node, create a join node
+            if (Graphs.vertexHasSuccessors(decomposition, decompNode)) {
+                decompNode = addJoin(decompNode).getFirst();
+            }
+
+            // calculate vertices of nearest successor, which needs to be handled
+            Set<V> clique = new HashSet<V>(predecessors);
+            clique.add(vertex);
+            Set<V> toIntroduce = new HashSet<V>(decompositionMap.get(decompNode));
+            toIntroduce.removeAll(clique);
```

Comment:
Or simply have `toIntroduce.removeAll(predecessors); toIntroduce.remove(vertex)`. Why do you need to create a new set?